### ISSUE TITLE
Add deprecation warning at runtime

### ISSUE THREAD

**Issue by mtrezza** [parse-community/parse-server#7342]
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation
The Deprecator currently only allows to detect deprecated Parse Server Options. It is not possible to manage deprecations that are only detectable at run-time.

### Feature / Enhancement Description
Extend the Deprecator to handle run-time deprecations.

It was expected that the deprecator will need to be extended to accommodate this scenario. The concept of the deprecator is to define deprecations centrally and compose the deprecation warning messages in a unified style, which is easy for Parse Server Options. For deprecations that are detected "in-code" or only at runtime we want to avoid spreading deprecation definitions all over the place that are difficult to manage, maintain and identify.

### Example Use Case
Log deprecation warning in https://github.com/parse-community/parse-server/pull/7339

### Alternatives / Workarounds
- Not phasing in run-time deprecations (violation of [Deprecation Policy](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#deprecation-policy))
- Write warning messages in-code (goes against the concept of adding deprecations in an easily manageable, unified way)

### 3rd Party References
n/a

**Comment by RaschidJFR** [comment]
Here's a sketchy idea to build on top of:

### Idea 1 (Runtime based)

1. Create a function `Deprecator.watchForDeprecatedKeys()` and use it like this:

    ```js
    function myFunction(params = { oldParam, newParam }) {

      // Check for use of the deprecated parameters
      Deprecator.watchForDeprecatedKeys(this, params, [{ 
        deprecatedKey: 'oldParam', 
        solution: 'Use newParam instead',
        willBeRemovedIn: '5.0'
      }]);

      // rest of the function's logic here...
    }
    ```

    **Behavior:**
    The function checks for the current version of `package.json`. If the version is higher than what set in param `willBeRemovedIn`, it will throw an exception like `"This feature should already be removed"`.

2. Add a test case for this feature (and each deprecated) in [Deprecator.spec.js](https://github.com/parse-community/parse-server/blob/master/spec/Deprecator.spec.js), so errors are caught by the CI:

    ```js
    describe('Watch for deprecated features', () => {
      it('someModule.myFunction', () => {
        spyOn(Deprecator, '_log');
        const { myFunction } = require('path/to/someModule');
        myFunction({ oldParam: 'I am deprecated!' });
        expect(Deprecator._log).toHaveBeenCalledWith(...);
      });
    });
    ```

### Idea 2 (build-time based)

Another alternative could be using preprocessing tools. For example, [jsdoc-api](https://www.npmjs.com/package/jsdoc-ap) would allow searching for *deprecated* keys in jsdoc comments; or maybe adding 'TODO' comments and linting with tools similar to [eslint-plugin-output-todo-comments](https://www.npmjs.com/package/eslint-plugin-output-todo-comments). Although, this ideas won't follow the intention of Deprecator.js.

**Comment by mtrezza** [comment]
I think 1 is the more pragmatic and fast to implement way. I'm still thinking about which aspects to standardize / for which deprecation scenarios to prepare, or if your example is already that's needed.

What we would not add is `willBeRemovedIn`, because a deprecation can be revoked or postponed, so we wouldn't want to suggest any date or version.

**Comment by RaschidJFR** [comment]
Yeah, option 1 is pretty straightforward, and also, it pseudo-centralizes deprecations if they are all added in `Deprecator.spec.js`. I can't think of a way of handling runtime warnings from `Deprecator.js` 🤔

So what's next? Need a hand with this? Should we wait for other ideas?

**Comment by mtrezza** [comment]
I'll make the PR in the coming days.

**Comment by mtrezza** [comment]
@RaschidJFR Quick update, I did not forget about this pending issue to merge your other PR. I will look into this shortly, apologies for the delay.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0](https://github.com/parse-community/parse-server/releases/tag/5.0.0)
